### PR TITLE
Add Ed25519, X25519 and small changes

### PR DESCRIPTION
**Pull request by rpiasetskyi**
- [ ] Ed25519
- [ ] X25519
- [ ] small changes

### CODE REVIEW

File: heimlig/src/crypto/ed25519.rs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+use ed25519_dalek::{SecretKey, Signature, Signer, SigningKey, Verifier, VerifyingKey};
+
+/// Ed25519 signature size in bytes.
+pub const SIGNATURE_SIZE: usize = ed25519_dalek::SIGNATURE_LENGTH;
+/// Ed25519 private key size in bytes.
+pub const PRIVATE_KEY_SIZE: usize = ed25519_dalek::SECRET_KEY_LENGTH;
+/// Ed25519 public key size in bytes.
+pub const PUBLIC_KEY_SIZE: usize = ed25519_dalek::PUBLIC_KEY_LENGTH;
+
+/// Signs the message using the Ed25519 algorithm.
```
Comment (inorick) [inline heimlig/src/crypto/ed25519.rs:11]: Good commenting style :+1:

File: heimlig/src/crypto/ed25519.rs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+///
+/// # Arguments
+///
+/// * `private_key`: A slice containing private key bytes.
+///   The private key has to be `PRIVATE_KEY_SIZE` bytes long.
+/// * `message`: A slice the message to sign bytes.
```
[the remark this answers is not in the document]
Comment (rpiasetskyi) [reply heimlig/src/crypto/ed25519.rs]: Done.
Thanks!

File: heimlig/src/crypto/mod.rs [hunk 3]
```diff
@@ -3,8 +3,10 @@ pub mod chacha20poly1305;
 mod ecc;
 pub mod ecdh;
 pub mod ecdsa;
+pub mod ed25519;
 pub mod hash;
 pub mod rng;
+pub mod x25519;
 
 /// Common errors.
 #[derive(Clone, Eq, PartialEq, Debug)]
```
[the remark this answers is not in the document]
Comment (rpiasetskyi) [reply heimlig/src/crypto/mod.rs:12]: Done.
https://github.com/eclipse-heimlig/heimlig/pull/158/commits/33e3181f39d9d130c577838d95e7d0395997aacf

File: heimlig/src/crypto/x25519.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    if slice.len() == KEY_SIZE {
+        Ok(())
+    } else {
+        Err(())
+    }
+}
+
+/// Converts slice to [u8; `KEY_SIZE`]. Returns an error if slice has wrong length.
+fn try_to_array(slice: &[u8]) -> Result<[u8; KEY_SIZE], ()> {
+    check_size(slice).map(|_| <[u8; KEY_SIZE]>::try_from(slice).unwrap())
```
Comment (inorick) [inline heimlig/src/crypto/x25519.rs:18]: Add comment why this `unwrap()` is ok.

File: heimlig/src/crypto/ed25519.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+/// * `signature`: A mutable slice where the computed signature will be stored.
+///   The signature slice length has to be `SIGNATURE_SIZE` bytes long.
+///
+/// # Errors
+///
+/// The function returns an error if:
+/// * `InvalidBufferSize`: The length of the `signature` is not `SIGNATURE_SIZE` bytes.
+/// * `InvalidPrivateKey`: The length of the `private_key` is not `PRIVATE_KEY_SIZE` bytes.
+pub fn ed25519_sign(private_key: &[u8], message: &[u8], signature: &mut [u8]) -> Result<(), Error> {
+    if signature.len() != SIGNATURE_SIZE {
```
Comment (inorick) [inline heimlig/src/crypto/ed25519.rs:27]: We should think about also allowing output buffers that are larger than required.

File: heimlig/src/crypto/x25519.rs (hunk 4 above)
Comment (rpiasetskyi) [reply heimlig/src/crypto/x25519.rs]: Done with `expect`.

File: heimlig/src/crypto/ed25519.rs (hunk 5 above)
Comment (rpiasetskyi) [reply heimlig/src/crypto/ed25519.rs:27]: It will be done for all algorithms in a separate commit.

**Pull request merged**